// react-native-keyboard-controller/lib/commonjs/animated.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.styles = exports.KeyboardProvider = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));

var _context = require("./context");

var _internal = require("./internal");

var _native = require("./native");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const KeyboardControllerViewAnimated = _reactNativeReanimated.default.createAnimatedComponent(_reactNative.Animated.createAnimatedComponent(_native.KeyboardControllerView));

const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1
  },
  hidden: {
    display: 'none',
    position: 'absolute'
  }
});

exports.styles = styles;

const KeyboardProvider = _ref => {
  let {
    children,
    statusBarTranslucent
  } = _ref;
  // animated values
  const progress = (0, _react.useRef)(new _reactNative.Animated.Value(0)).current;
  const height = (0, _react.useRef)(new _reactNative.Animated.Value(0)).current; // shared values

  const progressSV = (0, _reactNativeReanimated.useSharedValue)(0);
  const heightSV = (0, _reactNativeReanimated.useSharedValue)(0);
  const {
    setHandlers,
    broadcast
  } = (0, _internal.useSharedHandlers)(); // memo

  const context = (0, _react.useMemo)(() => ({
    animated: {
      progress: progress,
      height: _reactNative.Animated.multiply(height, -1)
    },
    reanimated: {
      progress: progressSV,
      height: heightSV
    },
    setHandlers
  }), []);
  const style = (0, _react.useMemo)(() => [styles.hidden, {
    transform: [{
      translateX: height
    }, {
      translateY: progress
    }]
  }], []);
  const onKeyboardMove = (0, _react.useMemo)(() => _reactNative.Animated.event([{
    nativeEvent: {
      progress,
      height
    }
  }], {
    useNativeDriver: true
  }), []); // handlers

  const updateSharedValues = (event, platforms) => {
    'worklet';

    if (platforms.includes(_reactNative.Platform.OS)) {
      progressSV.value = event.progress;
      heightSV.value = -event.height;
    }
  };

  const handler = (0, _internal.useAnimatedKeyboardHandler)({
    onKeyboardMoveStart: event => {
      'worklet';

      broadcast('onStart', event);
      updateSharedValues(event, ['ios']);
    },
    onKeyboardMove: event => {
      'worklet';

      broadcast('onMove', event);
      updateSharedValues(event, ['android']);
    },
    onKeyboardMoveEnd: event => {
      'worklet';

      broadcast('onEnd', event);
    }
  }, []);
  return /*#__PURE__*/_react.default.createElement(_context.KeyboardContext.Provider, {
    value: context
  }, /*#__PURE__*/_react.default.createElement(KeyboardControllerViewAnimated, {
    onKeyboardMoveReanimated: handler,
    onKeyboardMoveStart: _reactNative.Platform.OS === 'ios' ? onKeyboardMove : undefined,
    onKeyboardMove: _reactNative.Platform.OS === 'android' ? onKeyboardMove : undefined,
    statusBarTranslucent: statusBarTranslucent,
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
    // we are using this small hack, because if the component (where
    // animated value has been used) is unmounted, then animation will
    // stop receiving events (seems like it's react-native optimization).
    // So we need to keep a reference to the animated value, to keep it's
    // always mounted (keep a reference to an animated value).
    //
    // To test why it's needed, try to open screen which consumes Animated.Value
    // then close it and open it again (for example 'Animated transition').
    style: style
  }), children)));
};

exports.KeyboardProvider = KeyboardProvider;
